private/buf/cmd/buf/command/alpha/package/version/versionget/versionget.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package versionget
+
+import (
+	"context"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginref"
+	"github.com/bufbuild/buf/private/gen/proto/connect/buf/alpha/registry/v1alpha1/registryv1alpha1connect"
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/connectclient"
+	"github.com/bufbuild/connect-go"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+// NewCommand returns a new Command
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <buf.build/owner/repository[:ref]> <buf.build/owner/plugin[:version]>",
+		Short: "Resolve module and plugin reference to a specific remote package version",
+		Long: `This command returns the version of the asset to be used with one of the supported remote package registries.
+For example npm, go proxy, maven, swift.
+
+Examples:
+
+Get the version of the eliza module and the connect-go plugin for use with go modules.
+    $ buf alpha package version get buf.build/bufbuild/eliza buf.build/bufbuild/connect-go
+        v1.7.0-20230609151053-e682db0d9918.1
+
+Use a specific module version and plugin version.
+    $ buf alpha package version get buf.build/bufbuild/eliza:e682db0d99184be88b41c4405ea8a417 buf.build/bufbuild/connect-go:v1.7.0
+        v1.7.0-20230609151053-e682db0d9918.1
```

Comment:
I'm not sure that we ever concluded that the version will be in the same format across all possible language-specific registries, especially as num(languages) -> infinity. This may happen to be the case now, but I'm not sure we can guarantee this in the long-term. @mfridman thoughts on this?

Reply:
This is already true today, the maven registry has a [different format ](https://buf.build/docs/bsr/remote-packages/maven/#full-syntax), e.g., `0.1.7.1.20230403200515.aed131420688`

But we should be able to infer the versioning scheme based on the plugin. Every plugin is categorized, so if there is no client-side logic for building a version and it's returned from the server this should be okay?

Reply:
Is it true that every plugin MUST have exactly one language (and associated registry), or is this just a property of the current plugins we manage?

Reply:
all of this resolving is server side at the moment, if there's a plugin that isn't supported in a registry it will return an error. I figured we can change this behaviour later if we want it to default to something else

Reply:
If we promote this plugin to stable (or even beta), we can't later go in and split this command out on a per-registry basis. It feels like there's two options:

- Make plugins have at most 1 associated registry (which likely means language), i.e. make it so that a plugin either has no language, or 1 language. Examples: protoc-gen-go can only be used with the go module proxy, but a theoretical protoc-gen-doc would have no language (and no registry, more to the point).
- Split this command up into individual registries. Ie `buf package go-version`, `buf package maven-vesrsion`, `buf package npm-version`, where each errors if the plugin is not associated with the proper registry.

Reply:
AFAIK we've already got the concept of plugin -> registry type: [example](https://github.com/bufbuild/buf/blob/9e277a17d236f00016732cfdf4b012ce47b442f8/private/bufpkg/bufplugin/bufplugin.go#LL68C1-L68C1)

Plugins don't _need_ this type (because there's 
 [PLUGIN_REGISTRY_TYPE_UNSPECIFIED](https://github.com/bufbuild/buf/blob/9e277a17d236f00016732cfdf4b012ce47b442f8/proto/buf/alpha/registry/v1alpha1/plugin_curation.proto#L33)) it'll just mean that this command won't return anything if there isn't a registry attacked to it. 

I'm not a huge fan of option 2; it feels like it should be the servers responsibility to return a valid version if the plugins are valid. This also means that it's not dependent on peoples cli installation when we add more registries and the CLI will just act as a passthrough

Reply:
Option 1 seems reasonable. Worst case the server returns an "ambiguous" error to force the client to pass `--remote` or equivalent, but tbh I think this will be 1 plugin with at most 1 registry.

Reply:
Option 2 is saying that there are different formats for versions, and really different version types, which is likely more correct. You couldn't, for example, do `buf package version get | awk something`, as there are different layouts. What if I just want to extract the psuedo-timestamp? This is a property dependent on what the plugin is.

In reality, you're returning multiple potential objects from the same command. I think doing both option 1 and option 2 is likely the correct answer.

Reply:
What about returning a list of registry/version? In practice, I think most users will grep / awk or jq their way to the corresponding version and then extract the bits they need.

```
go       v1.7.0-20230609151053-e682db0d9918.1
maven    v0.1.7.1.20230403200515.aed131420688
``` 

or 

```json
{
  "go": {"version": "v1.7.0-20230609151053-e682db0d9918.1"},
  "maven": {"version": "v0.1.7.1.20230403200515.aed131420688"},
}
```

Having 1 command is less overhead for us and is a slightly nicer DX? To clarify, when you say "Split this command up into individual registries" .. do you mean to take the registry as the first arg or make each of the registries a sub-command?

Reply:
Make each of the registries a sub-command - can you explain the use case where a caller would not know the registry a plugin relates to at invocation time? ie when it is useful for this to be an effectively dynamically-typed command? My core contention is that `GoRegistryVersion` is a different type than `MavenRegistryVersion`, and this command is returning `any`.

Reply:
Ahhhh, okay I understand. I still think the client should print the version as-is without distinguishing because in practice a caller knows the plugin (and thus the registry) and will parse the bits based on the plugin they request.

> can you explain the use case where a caller would not know the registry a plugin relates to at invocation time

I cannot.

Reply:
Yea the client should print as-is, but if we can't think of a use case as to why they should be mixed, having separate commands seems more correct, and I'd argue actually provides better UX - I am declaring I know what type of version I want to get, and I know I only get that type of version back. We can also error if the plugin does not match the requested registry, ie a maven-version call for protoc-gen-go would error.

Reply:
That's fair, the only case I can see for slamming them together is making it easier for us to maintain and in the future when we add another registry a single command that takes plugin + module that returns `any` version will "just work" for older clients.

Reply:
If we intro new registries though, in theory no previously unsuccessful calls to this command will have been made that will now be newly successful. Even more to the point, I'd want protoc-gen-Haskell to error right now based on there being no valid registry for it, and it's weird for it to go from error to non-error just because we released a new feature (ie the Haskell registry). Tldr, I think separate commands is a less surprising overall experience for the user.

Reply:
should the API return the type of the registry through the already existing registryv1alpha1.PluginRegistryType, or should it populate a different field depending on what type the plugin is?
```
message GetRemotePackageVersionResponse {
    string go_version = 1; 
    string npm_version = 2;
    string maven_version = 3;
    string swift_version = 4;
}
```

Reply:
I'd imagine they'd be separate fields in a `oneof`, or likely even more correct is separate RPCs